aten/src/ATen/native/DispatchStub.cpp

```diff
@@ -95,6 +95,10 @@ void* DispatchStubImpl::get_call_ptr(
       TORCH_INTERNAL_ASSERT(hip_dispatch_ptr, "DispatchStub: missing HIP kernel");
       return hip_dispatch_ptr;
 
+    case DeviceType::VE:
+      TORCH_INTERNAL_ASSERT(ve_dispatch_ptr, "DispatchStub: missing VE kernel");
+      return ve_dispatch_ptr;
```

Comment:
You shouldn't need these DispatchStub changes. Did you actually exercise them in your extension?

Reply:
So far we had to use it as the ```at::copy``` for the HIP device relies on it. Further we still use it for ```copy_stub```, ```fill_stub```, ```mean_stub``` and ```sum_stub```.

But we can adapt it if necessary. As said before we maintain our integration since PyTorch v1.4.0 so we are used to changed APIs in every PyTorch release, i.e., v1.9.0 changes the argument order in the ```at::XXX.out``` functions.

Reply:
We used to have a number of inaccuracies in the dispatch table which would cause some calls on external backends to improperly route to DispatchStub. In 1.9.0 I believe that the DispatchStub is dead code; take Fill as an example:

```
- func: fill_.Scalar(Tensor(a!) self, Scalar value) -> Tensor(a!)
  device_check: NoCheck   # TensorIterator
  variants: function, method
  dispatch:
    CPU, CUDA, QuantizedCPU, QuantizedCUDA: fill_
    Meta: fill_meta_
```

The table now correctly says that `fill_` only works for CPU/CUDA backends, so you would have needed to directly register a `fill_` implementation.

I think I can believe that `copy_stub` is still being exercised since it doesn't look like we updated the dispatch table on it, but you should be able to override its behavior directly by registering to `copy_`. Can you check this?

Reply:
It works for ```fill_``` and ```copy_``` but it seems ```mean``` and ```sum``` in ReduceOps.cpp ( https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/ReduceOps.cpp#L788 ) still use stubs. I also get an ```RuntimeError: DispatchStub: unsupported device typeve``` thrown when I try to execute ```tensor.sum()``` although I registered: ```sum, sum.dim_IntList, sum.dim_DimnameList, sum.IntList_out, sum.DimnameList_out``` for the new device.

But at least we don't need the changes to ```ATen/native/Copy.cpp```.

Reply:
Just to confirm, this is testing on 1.9, is that right? (What hash?)

Reply:
Also, what did you register as the implementations?

Reply:
Yes. I tested it on the v1.9.0-rc3 tag as the master branch was not compiling.

The registration for the sum kernels looks like this:
```cpp
namespace ve {
	static at::Tensor sum(const at::Tensor& self, c10::optional<c10::ScalarType> dtype)
	static at::Tensor sum_dim_IntList(const at::Tensor& self, at::IntArrayRef dim, bool keepdim, c10::optional<at::ScalarType> opt_dtype)
	static at::Tensor sum_dim_DimnameList(const at::Tensor& self, at::DimnameList dim, bool keepdim, c10::optional<at::ScalarType> dtype)
	static at::Tensor& sum_DimnameList_out(const at::Tensor& self, at::DimnameList dim, bool keepdim, c10::optional<at::ScalarType> opt_dtype, at::Tensor& result)
	static at::Tensor& sum_IntList_out(const at::Tensor& self, at::IntArrayRef dim, bool keepdim, c10::optional<c10::ScalarType> opt_dtype, at::Tensor& result)
	static void sum_stub(at::TensorIterator& iter)
}

namespace at {
namespace native {
	REGISTER_VE_DISPATCH(sum_stub, &ve::sum_stub);
}}

TORCH_LIBRARY_IMPL(aten, VE, m) {
	m.impl("sum",			TORCH_FN(ve::sum));
	m.impl("sum.dim_IntList",	TORCH_FN(ve::sum_dim_IntList));
	m.impl("sum.dim_DimnameList",	TORCH_FN(ve::sum_dim_DimnameList));
	m.impl("sum.IntList_out",	TORCH_FN(ve::sum_IntList_out));
	m.impl("sum.DimnameList_out",	TORCH_FN(ve::sum_DimnameList_out));
}
```

Reply:
Hmph, this all looks correct. Any chance you can give a backtrace at the point `ve::sum_stub` is called? (e.g., gdb, breakpoint on ve::sum_stub, and the bt when it stops).

Reply:
I used an ```assert(false)``` instead of break point ;)

```
#0  0x00007ffff6c0f387 in raise () from /lib64/libc.so.6
#1  0x00007ffff6c10a78 in abort () from /lib64/libc.so.6
#2  0x00007ffff6c081a6 in __assert_fail_base () from /lib64/libc.so.6
#3  0x00007ffff6c08252 in __assert_fail () from /lib64/libc.so.6
#4  0x00007fffbc4eef58 in sum_stub (iter=...)
    at /.../op_mean.cpp:143
#5  0x00007ffec16beac6 in at::native::sum_out(at::Tensor const&, c10::ArrayRef<long>, bool, c10::optional<c10::ScalarType>, at::Tensor&) ()
   from /.../torch/lib/libtorch_cpu.so
#6  0x00007ffec16c02dd in at::native::sum(at::Tensor const&, c10::ArrayRef<long>, bool, c10::optional<c10::ScalarType>) ()
   from /.../torch/lib/libtorch_cpu.so
#7  0x00007fffbc4eea90 in sol::runtime::native::pytorch::ve::sum (self=..., dtype=...)
    at /.../op_mean.cpp:96
#8  0x00007fffbc4f1010 in c10::impl::detail::WrapFunctionIntoFunctor_<c10::CompileTimeFunctionPointer<at::Tensor(const at::Tensor&, c10::optional<c10::ScalarType>), sol::runtime::native::pytorch::ve::sum>, at::Tensor, c10::guts::typelist::typelist<const at::Tensor&, c10::optional<c10::ScalarType> > >::operator() (args#1=..., args#0=..., this=0x2fba730)
    at /.../torch/include/ATen/core/boxing/impl/WrapFunctionIntoFunctor.h:13
#9  c10::impl::wrap_kernel_functor_unboxed_<c10::impl::detail::WrapFunctionIntoFunctor_<c10::CompileTimeFunctionPointer<at::Tensor(const at::Tensor&, c10::optional<c10::ScalarType>), sol::runtime::native::pytorch::ve::sum>, at::Tensor, c10::guts::typelist::typelist<const at::Tensor&, c10::optional<c10::ScalarType> > >, at::Tensor(const at::Tensor&, c10::optional<c10::ScalarType>)>::call(c10::OperatorKernel *, c10::DispatchKeySet, const at::Tensor &, c10::optional<c10::ScalarType>) (functor=0x2fba730, args#0=..., args#1=...)
    at /.../torch/include/ATen/core/boxing/impl/make_boxed_from_unboxed_functor.h:423
#10 0x00007ffec1d20593 in at::redispatch::sum(c10::DispatchKeySet, at::Tensor const&, c10::optional<c10::ScalarType>) ()
   from /.../torch/lib/libtorch_cpu.so
#11 0x00007ffec33e047a in torch::autograd::VariableType::(anonymous namespace)::sum(c10::DispatchKeySet, at::Tensor const&, c10::optional<c10::ScalarType>) () from /.../torch/lib/libtorch_cpu.so
#12 0x00007ffec33e0632 in c10::impl::wrap_kernel_functor_unboxed_<c10::impl::detail::WrapFunctionIntoFunctor_<c10::CompileTimeFunctionPointer<at::Tensor (c10::DispatchKeySet, at::Tensor const&, c10::optional<c10::ScalarType>), &torch::autograd::VariableType::(anonymous namespace)::sum>, at::Tensor, c10::guts::typelist::typelist<c10::DispatchKeySet, at::Tensor const&, c10::optional<c10::ScalarType> > >, at::Tensor (c10::DispatchKeySet, at::Tensor const&, c10::optional<c10::ScalarType>)>::call(c10::OperatorKernel*, c10::DispatchKeySet, at::Tensor const&, c10::optional<c10::ScalarType>) () from /.../torch/lib/libtorch_cpu.so
#13 0x00007ffec2176009 in at::Tensor::sum(c10::optional<c10::ScalarType>) const ()
   from /.../torch/lib/libtorch_cpu.so
#14 0x00007fffdb9ba6f0 in torch::autograd::THPVariable_sum(_object*, _object*, _object*) ()
   from /.../torch/lib/libtorch_python.so
```

```op_mean.cpp:96``` is:
```cpp
static at::Tensor sum(const at::Tensor& self, c10::optional<c10::ScalarType> dtype) {
	return at::native::sum(self, std::vector<int64_t>{}, false, dtype);
}
```

Reply:
Thanks. The stub call is happening because you are routing through `native::sum` rather than `sum`. Are you using TensorIterator in your `sum_stub` implementation?

Reply:
Indeed, that was the problem 🤦🏻 Sorry. Yes we are using TensorIterator but we can create it on our own in the ```sum``` method.

I removed all stub related parts and was now able to run all tests without errors!

Reply:
Woohoo, thanks a lot!